doc/architecture.rst

```diff
@@ -143,16 +143,135 @@ the remote file will be downloaded and saved as message.txt.
+           "itemsperpage":""
+        },
+        "data":{
+           "version":{
+              "major":10,
+              "minor":0,
+              "micro":0,
+              "string":"10.0.0 beta",
+              "edition":"Community"
+           },
+           "capabilities":{
+              "core":{
+                 "pollinterval":60,
+                 "webdav-root":"remote.php/webdav"
+              },
+              "dav":{
+                 "chunking":"1.0"
+              },
+              "files_sharing":{
+                 "api_enabled":true,
+                 "public":{
+                    "enabled":true,
+                    "password":{
+                       "enforced":false
+                    },
+                    "expire_date":{
+                       "enabled":false
+                    },
+                    "send_mail":false,
+                    "upload":true
+                 },
+                 "user":{
+                    "send_mail":false
+                 },
+                 "resharing":true,
+                 "group_sharing":true,
+                 "federation":{
+                    "outgoing":true,
+                    "incoming":true
+                 }
+              },
+              "checksums":{
+                 "supportedTypes":[
+                    "SHA1"
+                 ],
+                 "preferredUploadType":"SHA1"
+              },
+              "files":{
+                 "bigfilechunking":true,
+                 "blacklisted_files":[
+                    ".htaccess"
+                 ],
+                 "undelete":true,
+                 "versioning":true
+              }
+           }
+        }
+     }
+  }
+
+Upload
+~~~~~~
+
+A checksum is calculated with the previously negotiated algorithm by the client and sent along with the file in an HTTP Header. 
+```OC-Checksum: [algorithm]:[checksum]```
+
+.. image:: ./images/checksums/client-activity.png
+
+During file upload, the server computes SHA1, MD5, and Adler-32 checksums and compares one of them to the checksum supplied by the client. 
+
+On mismatch, the server returns HTTP Status code 400 (Bad Request) thus signaling the client that the upload failed. 
+The server then discards the upload, and the client blacklists the file:
+
+.. image:: ./images/checksums/testing-checksums.png
+
+::
+
+  <?xml version='1.0' encoding='utf-8'?>
+  <d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
+    <s:exception>Sabre\DAV\Exception\BadRequest</s:exception>
+    <s:message>The computed checksum does not match the one received from the
+  client.</s:message>
+  </d:error>
+
+The client retries the upload using exponential back-off. 
+On success (matching checksum) the computed checksums are stored by the server in ``oc_filecache`` alongside the file.
+
+Chunked Upload
+~~~~~~~~~~~~~~
+
+Mostly same as above. 
+The checksum of the full file is sent with every chunk, but the server compares the checksum of the last chunk.
```

Comment:
I believe "The checksum of the full file is sent with every chunk" is correct.
It is just that when then reading further to "compares the checksum of the last chunk", it is ambiguous - because actually it is not "the checksum of the last chunk" but "the checksum sent with the last chunk".
Clear as mud???